File: services/chimera/src/commands/facts.rs
use async_trait::async_trait;
use duvua_framework::{
    builder::interaction_response::InteractionResponse,
    errors::BotError,
    handler::{CommandHandler, CommandHandlerData},
    utils::{get_option, get_sub_command},
};
use reqwest::Client;
use serenity::{
    builder::{CreateApplicationCommand, CreateApplicationCommandOption},
    model::prelude::{
        application_command::ApplicationCommandInteraction, command::CommandOptionType,
    },
    prelude::Context,
};

pub struct FactsCommand {
    client: Client,
    data: &'static CommandHandlerData,
}

impl FactsCommand {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            data: Box::leak(Box::new(build_data())),
        }
    }
}

#[async_trait]
impl CommandHandler for FactsCommand {
    async fn handle_command(
        &self,
        ctx: &Context,
        interaction: &ApplicationCommandInteraction,
    ) -> Result<(), BotError> {
        let sub_command = get_sub_command(&interaction.data.options)
            .ok_or(BotError::InvalidOption("sub-command"))?;

        let number = get_option(&sub_command.options, "number")
            .ok_or(BotError::OptionNotProvided("number"))?
            .value
            .ok_or(BotError::InvalidOption("number"))?
            .as_i64()
            .ok_or(BotError::InvalidOption("number"))?;

        let res = match sub_command.name.as_str() {
            "year" => text_req(&self.client, format!("http://numbersapi.com/{number}")).await?,
            "number" => text_req(&self.client, format!("http://numbersapi.com/{number}")).await?,
            _ => return Err(BotError::InvalidOption("sub-command")),
        };

        InteractionResponse::with_content(res.to_string())
            .respond(&ctx.http, interaction.id.0, &interaction.token)
            .await
    }

    fn get_data(&self) -> &'static CommandHandlerData {
        self.data
    }
}

async fn text_req(http: &Client, url: String) -> Result<String, BotError> {
    http.get(url)
        .send()
        .await
        .or_else(|e| {
            log::error!("Failed to make http request: {e}");
            Err(BotError::SomethingWentWrong)
        })?
        .text()
        .await
        .or_else(|e| {
            log::error!("Failed decode request body: {e}");
            Err(BotError::SomethingWentWrong)
        })
}

#[inline]
fn build_data() -> CommandHandlerData {
    CommandHandlerData {
        accepts_message_component: false,
        accepts_application_command: true,
        needs_defer: false,
        command_data: Some(build_data_command()),
        custom_id: None,
    }
}

fn build_data_command() -> CreateApplicationCommand {
    CreateApplicationCommand::default()
        .name("facts")
        .description("Exibe curiosidades sobre números")
        .description_localized("en-US", "Shows facts about numbers")
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("year")
                .description("Exibe curiosidades sobre um ano")
                .description_localized("en-US", "Show facts about some year")
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::Integer)
                        .name("number")
                        .description("O número que deseja pesquisar")
                        .description_localized("en-US", "The number you want to search")
                        .required(true)
                        .to_owned(),
                )
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("number")
                .description("Exibe curiosidades sobre um número qualquer")
                .description_localized("en-US", "Show facts about some arbitrary number")
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::Integer)
                        .name("number")
                        .description("O número que deseja pesquisar")
                        .description_localized("en-US", "The number you want to search")
                        .required(true)
                        .to_owned(),
                )
                .to_owned(),
        )
        .to_owned()
}

File: internal/commands/facts.go
package commands

import (
	"io"
	"mime"
	"net/http"
	"strconv"

	"github.com/bwmarrin/discordgo"
	"github.com/zanz1n/duvua-bot/internal/errors"
	"github.com/zanz1n/duvua-bot/internal/manager"
	"github.com/zanz1n/duvua-bot/internal/utils"
)

var factsNumberOpt = []*discordgo.ApplicationCommandOption{
	{
		Type:        discordgo.ApplicationCommandOptionInteger,
		Name:        "number",
		Description: "O número que deseja pesquisar (caso não informado será aleatório)",
		DescriptionLocalizations: map[discordgo.Locale]string{
			discordgo.EnglishUS: "The number you want to search (if not set will be random)",
		},
	},
}

var factsCommandData = discordgo.ApplicationCommand{
	Name:        "facts",
	Type:        discordgo.ChatApplicationCommand,
	Description: "Exibe curiosidades sobre números",
	DescriptionLocalizations: &map[discordgo.Locale]string{
		discordgo.EnglishUS: "Shows facts about numbers (if not set will be random)",
	},
	Options: []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "year",
			Description: "Exibe curiosidades sobre um ano",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Show facts about some year",
			},
			Options: factsNumberOpt,
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "number",
			Description: "Exibe curiosidades sobre um número qualquer",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Show facts about some arbitrary number",
			},
			Options: factsNumberOpt,
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "math",
			Description: "Exibe curiosidades de matemática sobre um número qualquer",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Show math facts about some arbitrary number",
			},
			Options: factsNumberOpt,
		},
	},
}

func NewFactsCommand() manager.Command {
	return manager.Command{
		Accepts: manager.CommandAccept{
			Slash:  true,
			Button: false,
		},
		Data:       &factsCommandData,
		Category:   manager.CommandCategoryInfo,
		NeedsDefer: false,
		Handler:    &FactsCommand{},
	}
}

type FactsCommand struct {
}

func (c *FactsCommand) Handle(s *discordgo.Session, i *discordgo.InteractionCreate) error {
	if i.Type != discordgo.InteractionApplicationCommand &&
		i.Type != discordgo.InteractionApplicationCommandAutocomplete {
		return errors.New("interação de tipo inesperado")
	}

	data := i.ApplicationCommandData()

	subCommand := utils.GetSubCommand(data.Options)
	if subCommand == nil {
		return errors.New("opção `sub-command` é necessária")
	}

	number := "random"
	if numberOpt := utils.GetOption(subCommand.Options, "number"); numberOpt != nil {
		if numberOpt.Type != discordgo.ApplicationCommandOptionInteger {
			return errors.New("opção `amount` precisa ser um número inteiro")
		}
		number = strconv.FormatInt(numberOpt.IntValue(), 10)
	}

	const baseUrl = "http://numbersapi.com/"
	var url string
	switch subCommand.Name {
	case "year":
		url = baseUrl + number + "/" + "year"
	case "math":
		url = baseUrl + number + "/" + "math"
	default:
		url = baseUrl + number + "/" + "trivia"
	}

	res, err := s.Client.Get(url)
	if err != nil {
		return errors.Unexpected("failed to fetch numbers api: " + err.Error())
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		return errors.Unexpectedf(
			"failed to fetch numbers api: status code %v",
			res.StatusCode,
		)
	} else if mt, _, _ := mime.ParseMediaType(res.Header.Get("Content-Type")); mt != "text/plain" {
		return errors.Unexpectedf(
			"failed to fetch numbers api: unexpected content type %s",
			mt,
		)
	}

	bodyBuf, err := io.ReadAll(res.Body)
	if err != nil {
		return errors.Unexpected("failed to fetch numbers api: " + err.Error())
	}
	bodyS := string(bodyBuf)

	return s.InteractionRespond(i.Interaction, utils.BasicResponse(bodyS))
}
